### PR TITLE
Data/series-interpolate

### PR DESCRIPTION
**Pull request by macflo8**
This PR is an attempt at refactoring `interpolate()` to use `_data`. Tests with `extra_cols` do not work yet

### CODE REVIEW

File: pyam/core.py [hunk 1]
```diff
@@ -382,17 +382,32 @@ def interpolate(self, time):
...
+                for i in range (len(year)):
+                    if year[i] > interpolate_at:
+                        if i != 0:
+                            p = year[i-1]
+                            n = year[i]
+                            value = (((n - interpolate_at) * x[i-1]
+                                + (interpolate_at - p) * x[i]) / (n - p))
+                            break
+                new_key = a+(interpolate_at,)
+                print(new_key)
```
Comment (danielhuppmann) [inline pyam/core.py:407]: remove left-over print statement
```suggestion
                new_key = a+(interpolate_at,)
```

File: pyam/core.py [hunk 2]
```diff
@@ -382,17 +382,32 @@ def interpolate(self, time):
...
+
+        for a, key in grouped:
+            year = [y for y in key.index.get_level_values('year')]
+            if (year[-1] > interpolate_at and year[0] < interpolate_at 
+            and interpolate_at not in year):
+                x = [x for x in key]
+                for i in range (len(year)):
+                    if year[i] > interpolate_at:
+                        if i != 0:
+                            p = year[i-1]
```
Comment (danielhuppmann) [inline pyam/core.py:401]: this section seems to re-implement the `fill_values()` function...?

File: pyam/core.py [hunk 3]
```diff
@@ -382,17 +382,32 @@ def interpolate(self, time):
...
-                              values='value', aggfunc=np.sum)
-        # drop time-rows where values are already defined
-        if time in df.columns:
-            df = df[np.isnan(df[time])]
-        fill_values = df.apply(fill_series, raw=False, axis=1, time=time)
-        fill_values = fill_values.dropna().reset_index()
-        fill_values = fill_values.rename(columns={0: "value"})
-        fill_values[self.time_col] = time
-        self.data = self.data.append(fill_values, ignore_index=True)
+        grouped = self._data.groupby(level=index)
```
Comment (danielhuppmann) [inline pyam/core.py:387]: this seems to be a duplication of the line below

File: pyam/core.py [hunk 4]
```diff
@@ -382,17 +382,32 @@ def interpolate(self, time):
...
+        interpolate_at = time
```
Comment (danielhuppmann) [inline pyam/core.py:388]: why is this rename necessary?

File: pyam/core.py [hunk 5]
```diff
@@ -382,17 +382,32 @@ def interpolate(self, time):
...
+        multi_index_dict = {}
+
+        grouped = self._data.groupby(level=index)
+
+        for a, key in grouped:
```
Comment (danielhuppmann) [inline pyam/core.py:393]: The variables `a` and `key` are not very helpful - it seems that the first arg is the group-index and the second is the data (as pd.Series), so `key` and `values` would be more logical names.

File: pyam/core.py (hunk 3 above)
Comment (danielhuppmann) [inline pyam/core.py:387]: This seems to be a duplication of the line below

File: pyam/core.py (hunk 4 above)
Comment (danielhuppmann) [inline pyam/core.py:388]: why is this renaming necessary?

**Comment by macflo8** [comment]
closing in favour of #25
